File: app/src/main/java/com/battlelancer/seriesguide/backend/RemoveCloudAccountDialogFragment.java
package com.battlelancer.seriesguide.backend;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatDialogFragment;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.util.Errors;
import com.battlelancer.seriesguide.util.Utils;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.uwetrottmann.seriesguide.backend.account.Account;
import java.io.IOException;
import org.greenrobot.eventbus.EventBus;

/**
 * Confirms whether to obliterate a SeriesGuide cloud account. If removal is tried, posts result as
 * {@link AccountRemovedEvent}. If dialog is canceled, posts a {@link CanceledEvent}.
 */
public class RemoveCloudAccountDialogFragment extends AppCompatDialogFragment {

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setMessage(R.string.hexagon_remove_account_confirmation);
        builder.setPositiveButton(R.string.hexagon_remove_account,
                (dialog, which) -> Utils.executeInOrder(new RemoveHexagonAccountTask(getContext()))
        );
        builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> sendCanceledEvent());

        return builder.create();
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        super.onCancel(dialog);
        sendCanceledEvent();
    }

    private void sendCanceledEvent() {
        EventBus.getDefault().post(new CanceledEvent());
    }

    public static class RemoveHexagonAccountTask extends AsyncTask<Void, Void, Boolean> {

        @SuppressLint("StaticFieldLeak") private final Context context;
        private final HexagonTools hexagonTools;

        public RemoveHexagonAccountTask(Context context) {
            this.context = context.getApplicationContext();
            this.hexagonTools = SgApp.getServicesComponent(context).hexagonTools();
        }

        @Override
        protected Boolean doInBackground(Void... params) {
            // remove account from hexagon
            try {
                Account accountService = hexagonTools.buildAccountService();
                if (accountService == null) {
                    return false;
                }
                accountService.deleteData().execute();
            } catch (IOException e) {
                Errors.logAndReportHexagon("remove account", e);
                return false;
            }

            // de-authorize app so other clients are signed out as well
            GoogleSignInClient googleSignInClient = GoogleSignIn
                    .getClient(context, HexagonTools.getGoogleSignInOptions());
            Task<Void> task = googleSignInClient.revokeAccess();
            try {
                Tasks.await(task);
            } catch (Exception e) {
                Errors.logAndReport("remove account", HexagonAuthError.build("remove account", e));
                return false;
            }

            // disable Hexagon integration, remove local account data
            hexagonTools.setDisabled();

            return true;
        }

        @Override
        protected void onPostExecute(Boolean result) {
            EventBus.getDefault().post(new AccountRemovedEvent(result));
        }
    }

    public static class CanceledEvent {
    }

    public static class AccountRemovedEvent {
        public final boolean successful;

        public AccountRemovedEvent(boolean successful) {
            this.successful = successful;
        }

        /**
         * Display status toasts depending on the result.
         */
        public void handle(Context context) {
            Toast.makeText(context, successful ? R.string.hexagon_remove_account_success
                    : R.string.hexagon_remove_account_failure, Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/backend/RemoveCloudAccountDialogFragment.kt
// Copyright 2023 Uwe Trottmann
// SPDX-License-Identifier: Apache-2.0

package com.battlelancer.seriesguide.backend

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatDialogFragment
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.SgApp.Companion.getServicesComponent
import com.battlelancer.seriesguide.backend.RemoveCloudAccountDialogFragment.AccountRemovedEvent
import com.battlelancer.seriesguide.backend.RemoveCloudAccountDialogFragment.CanceledEvent
import com.battlelancer.seriesguide.util.Errors
import com.firebase.ui.auth.AuthUI
import com.google.android.gms.tasks.Tasks
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.withContext
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.ExecutionException

/**
 * Confirms whether to obliterate a SeriesGuide cloud account. If removal is tried, posts result as
 * [AccountRemovedEvent]. If dialog is canceled, posts a [CanceledEvent].
 */
class RemoveCloudAccountDialogFragment : AppCompatDialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return MaterialAlertDialogBuilder(requireContext())
            .setMessage(R.string.hexagon_remove_account_confirmation)
            .setPositiveButton(R.string.hexagon_remove_account) { _: DialogInterface?, _: Int ->
                SgApp.coroutineScope.launch {
                    RemoveHexagonAccountTask(requireContext()).run()
                }
            }
            .setNegativeButton(android.R.string.cancel) { _: DialogInterface?, _: Int ->
                sendCanceledEvent()
            }
            .create()
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)
        sendCanceledEvent()
    }

    private fun sendCanceledEvent() {
        EventBus.getDefault().post(CanceledEvent())
    }

    class RemoveHexagonAccountTask(context: Context) {

        private val context: Context = context.applicationContext
        private val hexagonTools: HexagonTools = getServicesComponent(context).hexagonTools()

        suspend fun run() {
            withContext(Dispatchers.IO) {
                removeJobSemaphore.withPermit {
                    val result = doInBackground()
                    onPostExecute(result)
                }
            }
        }

        private fun doInBackground(): Boolean {
            // remove account from hexagon
            try {
                val accountService = hexagonTools.buildAccountService()
                    ?: return false
                accountService.deleteData().execute()
            } catch (e: IOException) {
                Errors.logAndReportHexagon(ACTION_REMOVE_ACCOUNT, e)
                return false
            }

            // Delete Firebase account so other clients are signed out as well
            val task = AuthUI.getInstance().delete(context)
            try {
                Tasks.await(task)
            } catch (e: Exception) {
                // https://developers.google.com/android/reference/com/google/android/gms/tasks/Tasks#public-static-tresult-await-tasktresult-task
                if (e is InterruptedException) {
                    // Do not report thread interruptions, it's expected.
                    Timber.w(e, ACTION_REMOVE_ACCOUNT)
                } else {
                    val cause = if (e is ExecutionException) {
                        e.cause ?: e // The Task failed, getCause returns the original exception.
                    } else {
                        e // Unexpected exception.
                    }
                    val authEx = HexagonAuthError.build(ACTION_REMOVE_ACCOUNT, cause)
                    Errors.logAndReportHexagonAuthError(authEx)
                }
                return false
            }

            // disable Hexagon integration, remove local account data
            hexagonTools.removeAccountAndSetDisabled()
            return true
        }

        private fun onPostExecute(result: Boolean) {
            EventBus.getDefault().post(AccountRemovedEvent(result))
        }

        companion object {
            private const val ACTION_REMOVE_ACCOUNT = "remove account"
            private val removeJobSemaphore = Semaphore(permits = 1)
        }
    }

    class CanceledEvent

    class AccountRemovedEvent(val successful: Boolean) {
        /**
         * Display status toasts depending on the result.
         */
        fun handle(context: Context) {
            Toast.makeText(
                context,
                if (successful) R.string.hexagon_remove_account_success else R.string.hexagon_remove_account_failure,
                Toast.LENGTH_LONG
            ).show()
        }
    }
}